src/Features/Core/Portable/NavigateTo/AbstractNavigateToSearchService.InProcess.cs

```diff
@@ -268,6 +268,31 @@ private static NavigateToMatchKind GetNavigateToMatchKind(ArrayBuilder<PatternMa
                 return NavigateToMatchKind.Substring;
             }
 
+            if (nameMatches.Any(r => r.Kind == PatternMatchKind.CamelCaseExact))
+            {
+                return NavigateToMatchKind.CamelCaseExact;
+            }
+
+            if (nameMatches.Any(r => r.Kind == PatternMatchKind.CamelCasePrefix))
+            {
+                return NavigateToMatchKind.CamelCasePrefix;
+            }
+
+            if (nameMatches.Any(r => r.Kind == PatternMatchKind.CamelCaseNonContiguousPrefix))
+            {
+                return NavigateToMatchKind.CamelCaseNonContiguousPrefix;
+            }
+
+            if (nameMatches.Any(r => r.Kind == PatternMatchKind.CamelCaseSubstring))
+            {
+                return NavigateToMatchKind.CamelCaseSubstring;
+            }
+
+            if (nameMatches.Any(r => r.Kind == PatternMatchKind.CamelCaseNonContiguousSubstring))
+            {
+                return NavigateToMatchKind.CamelCaseNonContiguousSubstring;
+            }
```

Comment:
okay. I am a bit confused. so we have 3 almost exactly same enum

PatternMatchKind from CodeAnalysis, PatternMatchKind from editor and NavigateToMatchKind from CodeAnalysis.

all these are basically exactly same. NavigateToMatchKind has some extra flag but those are not used anywhere.

I can understand Roslyn has its own enum since we are rehostable  and should work without VS dependency. but not sure why we have both exact same NavigateToMatchKind and PatternMatchKind. can we just use PatternMatchKind everywhere and remove NavigateToMatchKind?

Reply:
as Oleg stated above, INavigateToSearchResult is used by TypeScript thanks to InternalsVisibleTo. INavigateToSearchResult is also used to pass search results from portable to VS-dependent layers and it cannot reference editor's PatternMatchKind because editor API is not yet portable.
The solution we came up with was to "extend" NavigateToMatchKind enum so that Roslyn can preserve pattern match kind information, but still not break TypeScript.

Reply:
I am not talking about editor's PatternMatchKind. I am talking about our PatternMatchKind. and for typescript. I think it is completely fine to ask them to change to use PatternmatchKind since we are removing NavigateTomatchKind.

Reply:
How would you change INavigateToSearchResult without breaking TypeScript? They are in a different repo. INavigateToSearchResult2? That's what we do in the editor typically.

Also I'd suggest not exposing Roslyn's PatternMatchKind in "internal-but-visible-to-partner-teams" API as going forward I still hope we can make editor definitions portable and then Roslyn can delete its copy of the pattern matcher in favor of editor's one.

Reply:
typescript is already using internal type through internal visible to. so there is no reason we keep the backward compatibility like it is public type. and I dont think that conflict with 
> "as going forward I still hope we can make editor definitions portable and then Roslyn can delete its copy of the pattern matcher in favor of editor's one."

CodeAnalysis.PatternMatchKind or CodeAnalysis.NavigateToMatchKind, it is just name. both are still internal enums.

Reply:
> typescript is already using internal type through internal visible to. so there is no reason we keep the backward compatibility like it is public type. and I dont think that conflict with

We need backcompat *for now* until we can coordinate with TS and F# to make changes to these APIs we share with them.